triggers/tests/src/test/scala/com/daml/trigger/test/TestMain.scala

```diff
@@ -0,0 +1,421 @@
+
+import com.daml.trigger.{Runner, TriggerMsg}
+
+case class Config(ledgerPort: Int, darPath: File)
+
+// We do not use scalatest here since that doesn’t work nicely with
+// the client_server_test macro.
+
+case class NumMessages(num: Long)
+
+object TestRunner {
+  def assertEqual[A](actual: A, expected: A) = {
+    if (actual == expected) {
+      Right(())
+    } else {
+      Left(s"Expected $actual but got $expected")
+    }
+  }
+  def findStdlibPackageId(dar: Dar[(PackageId, Package)]) =
+    dar.all
+      .find {
+        case (pkgId, pkg) =>
+          pkg.modules.contains(DottedName.assertFromString("DA.Internal.Template"))
+      }
+      .get
+      ._1
+}
+
+class TestRunner(ledgerPort: Int) {
+  var partyCount = 0
+
+  val applicationId = ApplicationId("Trigger Test Runner")
+
+  val clientConfig = LedgerClientConfiguration(
+    applicationId = applicationId.unwrap,
+    ledgerIdRequirement = LedgerIdRequirement("", enabled = false),
+    commandClient = CommandClientConfiguration.default,
+    sslContext = None
+  )
+
+  def getNewParty(): String = {
+    partyCount = partyCount + 1
+    s"Alice$partyCount"
+  }
+
+  def genericTest[A](
+      // test name
+      name: String,
+      // The dar package the trigger is in
+      dar: Dar[(PackageId, Package)],
+      // Identifier of the trigger value
+      triggerId: Identifier,
+      // Commands to interact with the trigger
+      commands: (LedgerClient, String) => ExecutionContext => ActorMaterializer => Future[A],
+      // the number of messages that should be delivered to the trigger
+      numMessages: NumMessages,
+      // assertion on final state
+      assertFinalState: (SExpr, A) => Either[String, Unit],
+      // assertion on the final acs represented as a map from template_id to
+      // (contract_id, create_arguments)
+      assertFinalACS: (
+          Map[Identifier, Seq[(String, Lf.ValueRecord[Lf.AbsoluteContractId])]],
+          A) => Either[String, Unit]) = {
+
+    println(s"---\n$name:")
+
+    val system = ActorSystem("TriggerRunner")
+    val sequencer = new AkkaExecutionSequencerPool("TriggerRunnerPool")(system)
+    implicit val materializer: ActorMaterializer = ActorMaterializer()(system)
+    implicit val ec: ExecutionContext = system.dispatcher
+
+    val party = getNewParty()
+    val clientF = LedgerClient.singleHost("localhost", ledgerPort, clientConfig)(ec, sequencer)
+    val triggerFlow: Future[SExpr] = for {
+      client <- clientF
+      finalState <- Runner.run(
+        dar,
+        triggerId,
+        client,
+        applicationId,
+        party,
+        msgFlow = Flow[TriggerMsg].take(numMessages.num)
+      )
+    } yield finalState
+    val commandsFlow: Future[A] =
+      clientF.flatMap(client => commands(client, party)(ec)(materializer))
+    triggerFlow.failed.foreach(_ => system.terminate)
+    commandsFlow.failed.foreach(_ => system.terminate)
```

Comment:
Should we log the errors here to make failures easier to debug?

Reply:
I believe we already log those error in `testFlow.onComplete`.